lua/apisix/plugins/batch-processor.lua

```diff
@@ -0,0 +1,179 @@
+local setmetatable = setmetatable
+local timer_at = ngx.timer.at
+local remove = table.remove
+local fmt = string.format
+local ngx_log = ngx.log
+local DEBUG = ngx.DEBUG
+local table = table
+local now = ngx.now
+local type = type
+local Batch_Processor = {}
+local Batch_Processor_mt = {
+    __index = Batch_Processor
+}
+local execute_func
+local create_buffer_timer
+
+
+local schema = {
+    type = "object",
+    properties = {
+        name = {type = "string", default = "log buffer"},
+        max_retry_count = {type = "integer", minimum = 0, default= 0},
+        retry_delay = {type = "integer", minimum = 0, default= 1},
+        buffer_duration = {type = "integer", minimum = 1, default= 60}, -- maximum age in seconds of the oldest log item in a batch before the batch must be transmitted
+        inactive_timeout = {type = "integer", minimum = 1, default= 5}, -- maximum age in seconds when the buffer will be flushed if inactive
+        batch_max_size = {type = "integer", minimum = 1, default= 1000}, -- maximum number of entries in a batch before the batch must be transmitted
+    }
+}
+
+
+local function schedule_func_exec(batch_processor, delay, batch)
+    local hdl, err = timer_at(delay, execute_func, batch_processor, batch)
+    if not hdl then
+        core.log.error("failed to create process timer: ", err)
+        return
+    end
+end
+
+
+execute_func = function(premature, batch_processor, batch)
+    if premature then
+        return
+    end
+
+    local ok, err = batch_processor.func(batch.entries)
+    if ok then
+        ngx_log(DEBUG, fmt("Batch Processor[%s] successfully processed the entries", batch_processor.name))
+
+    else
+        batch.retry_count = batch.retry_count + 1
+        if batch.retry_count < batch_processor.max_retry_count then
+            core.log.warn(fmt("Batch Processor[%s] failed to process entries: ", batch_processor.name), err)
+            schedule_func_exec(batch_processor, batch_processor.retry_delay, batch)
+        else
+            core.log.error(fmt("Batch Processor[%s] exceeded the max_retry_count[%d], dropping the entries", batch_processor.name, batch.retry_count))
+        end
+    end
+end
+
+
+local function flush_buffer(premature, batch_processor)
+
+    if premature then
+        return
+    end
+
+    if now() - batch_processor.last_entry_t >= batch_processor.inactive_timeout or
+            now() - batch_processor.first_entry_t >= batch_processor.buffer_duration then
+        ngx_log(DEBUG, fmt("BatchProcessor[%s] buffer duration exceeded, activating buffer flush", batch_processor.name))
+        batch_processor:process_buffer()
+        batch_processor.isTimerRunning = false
+        return
+    end
+
+    -- buffer duration did not exceed or the buffer is active, extending the timer
+    ngx_log(DEBUG, fmt("BatchProcessor[%s] extending buffer timer", batch_processor.name))
+    create_buffer_timer(batch_processor)
+end
+
+
+create_buffer_timer = function(batch_processor)
```

Comment:
This is used as a forward declaration in the local variable( line 32). Because the create buffer timer is referenced in a previous method (line 93) to delay the timer if there is any activity found in the buffer.

Reply:
good style for this case:  `function create_buffer_timer(batch_processor)`

Reply:
same problem with function `execute_func`

Reply:
Resolved